File: set3/8/main.py
import sys

def input():
    return sys.stdin.readline().rstrip()

N, K = map(int, input().split())

go        = [-1] * (N + 1)
idx       = [0] * (N + 1)
arr       = [0] * (N + 1)
tmp       = [0] * (N + 1)
index_arr = [0] * (N + 1)
numbering = [0] * (N + 1)
V         = [[] for i in range(N + 1)]

arr[1:N+1] = list(map(int, input().split()))
idx[1:N+1] = list(map(int, input().split()))

for i in range(1, N + 1):
    if go[i] == -1:
        cur, number = i, 1
        while True:
            go[cur] = number
            index_arr[cur] = i 
            V[i].append(cur)
            if go[idx[cur]] != -1: break
            cur = idx[cur]
            number += 1
    siz  = len(V[index_arr[i]])
    T    = K % siz
    Vidx = (go[i] + T - 1) % siz
    pre  = V[index_arr[i]][Vidx]
    tmp[pre] = arr[i]
print(" ".join(map(str, tmp[1:N+1])))

File: set3/8/main.cpp
#include<iostream>
#include<vector>

using namespace std;

vector<int> go, idx, arr, tmp;
vector<int> index_arr, numbering;
vector<vector<int>> V;

void dfs(int cur, int root, int number) {
    go[cur] = number;
    index_arr[cur] = root;
    V[root].push_back(cur);
    if(go[idx[cur]] == -1) dfs(idx[cur], root, number + 1);
}

int main(){
    ios::sync_with_stdio(false);
    cin.tie(0);

    int N; long long K; cin >> N >> K;
    go = vector<int>(N + 1, -1);
    index_arr = arr = tmp = idx = vector<int>(N + 1);
    V.resize(N + 1);
    for(int i = 1; i <= N; i++) cin >> arr[i];
    for(int i = 1; i <= N; i++) cin >> idx[i];
    
    for(int i = 1; i <= N; i++) {
        if(go[i] == -1) dfs(i, i, 1);
        int idx  = index_arr[i];
        int siz  = V[idx].size();
        int T    = K % siz;
        int Vidx = (go[i] + T - 1) % siz;
        int pre  = V[idx][Vidx];
        tmp[pre] = arr[i];
    }

    arr = tmp;
    for(int i = 1; i <= N; i++) cout << arr[i] << " ";

    return 0;
}